```
self = <AsyncMock name='_run_cephadm' id='136408502725088'>
args = ('test', 'nvmeof.testpool.mygroup.test.qwert', ['_orch', 'deploy'], [])
kwargs = {'error_ok': True, 'stdin': '{"fsid": "fsid", "name": "nvmeof.testpool.mygroup.test.qwert", "image": "", "deploy_argum...e_in_usecs = 0\\nnotifications_interval = 60\\n\\n[monitor]\\ntimeout = 1.0\\n"}}}', 'use_current_daemon_image': False}
expected = call('test', 'nvmeof.testpool.mygroup.test.qwert', ['_orch', 'deploy'], [], stdin='{"fsid": "fsid", "name": "nvmeof.te... 0\\nnotifications_interval = 60\\n\\n[monitor]\\ntimeout = 1.0\\n"}}}', error_ok=True, use_current_daemon_image=False)
actual = call('test', 'nvmeof.testpool.mygroup.test.qwert', ['_orch', 'deploy'], [], stdin='{"fsid": "fsid", "name": "nvmeof.te... 0\\nnotifications_interval = 60\\n\\n[monitor]\\ntimeout = 1.0\\n"}}}', use_current_daemon_image=False, error_ok=True)
_error_message = <function NonCallableMock.assert_called_with.<locals>._error_message at 0x7c103421cca0>
cause = None

    def assert_called_with(self, /, *args, **kwargs):
        """assert that the last call was made with the specified arguments.
    
        Raises an AssertionError if the args and keyword args passed in are
        different to the last call to the mock."""
        if self.call_args is None:
            expected = self._format_mock_call_signature(args, kwargs)
            actual = 'not called.'
            error_message = ('expected call not found.\nExpected: %s\nActual: %s'
                    % (expected, actual))
            raise AssertionError(error_message)
    
        def _error_message():
            msg = self._format_mock_failure_message(args, kwargs)
            return msg
        expected = self._call_matcher(_Call((args, kwargs), two=True))
        actual = self._call_matcher(self.call_args)
        if actual != expected:
            cause = expected if isinstance(expected, Exception) else None
>           raise AssertionError(_error_message()) from cause
E           AssertionError: expected call not found.
E           Expected: _run_cephadm('test', 'nvmeof.testpool.mygroup.test.qwert', ['_orch', 'deploy'], [], stdin='{"fsid": "fsid", "name": "nvmeof.testpool.mygroup.test.qwert", "image": "", "deploy_arguments": [], "params": {"tcp_ports": [5500, 4420, 8009, 10008]}, "meta": {"service_name": "nvmeof.testpool.mygroup", "ports": [5500, 4420, 8009, 10008], "ip": null, "deployed_by": [], "rank": null, "rank_generation": null, "extra_container_args": null, "extra_entrypoint_args": null}, "config_blobs": {"config": "", "keyring": "[client.nvmeof.testpool.mygroup.test.qwert]\\nkey = None\\n", "files": {"ceph-nvmeof.conf": "# This file is generated by cephadm.\\n[gateway]\\nname = client.nvmeof.testpool.mygroup.test.qwert\\ngroup = mygroup\\naddr = 192.168.100.100\\nport = 5500\\nenable_auth = False\\nstate_update_notify = True\\nstate_update_interval_sec = 5\\nbreak_update_interval_sec = 25\\nenable_spdk_discovery_controller = False\\nencryption_key = /encryption.key\\nrebalance_period_sec = 7\\nmax_gws_in_grp = 16\\nmax_ns_to_change_lb_grp = 8\\nenable_prometheus_exporter = True\\nprometheus_exporter_ssl = False\\nprometheus_port = 10008\\nprometheus_stats_interval = 10\\nprometheus_startup_delay = 240\\nprometheus_connection_list_cache_expiration = 60\\nverify_nqns = True\\nverify_keys = True\\nverify_listener_ip = True\\n# This is a development flag, do not change it\\nabort_on_errors = True\\n# This is a development flag, do not change it\\nabort_on_update_error = True\\n# This is a development flag, do not change it\\nomap_file_ignore_unlock_errors = False\\n# This is a development flag, do not change it\\nomap_file_lock_on_read = True\\nomap_file_lock_duration = 20\\nomap_file_lock_retries = 30\\nomap_file_lock_retry_sleep_interval = 1.0\\nomap_file_update_reloads = 10\\nallowed_consecutive_spdk_ping_failures = 1\\nspdk_ping_interval_in_seconds = 2.0\\nping_spdk_under_lock = False\\nenable_monitor_client = True\\nmax_hosts_per_namespace = 16\\nmax_namespaces_with_netmask = 1000\\nmax_subsystems = 128\\nmax_hosts = 2048\\nmax_namespaces = 4096\\nmax_namespaces_per_subsystem = 512\\nmax_hosts_per_subsystem = 128\\nsubsystem_cache_expiration = 5\\nforce_tls = False\\n# This is a development flag, do not change it\\nmax_message_length_in_mb = 4\\n\\n[gateway-logs]\\nlog_level = INFO\\nlog_files_enabled = True\\nlog_files_rotation_enabled = True\\nverbose_log_messages = True\\nmax_log_file_size_in_mb = 10\\nmax_log_files_count = 20\\nmax_log_directory_backups = 10\\nlog_directory = /var/log/ceph/\\n\\n[discovery]\\naddr = 192.168.100.100\\nport = 8009\\n# This is a development flag, do not change it\\nabort_on_errors = True\\nbind_retries_limit = 10\\nbind_sleep_interval = 0.5\\n\\n[ceph]\\npool = testpool\\nconfig_file = /etc/ceph/ceph.conf\\nid = nvmeof.testpool.mygroup.test.qwert\\n\\n[mtls]\\nserver_key = /server.key\\nclient_key = /client.key\\nserver_cert = /server.cert\\nclient_cert = /client.cert\\nroot_ca_cert = /root.ca.cert\\n\\n[kmip]\\ncert_dir = ./certs/kmip/{server_name}\\n\\n[spdk]\\ntgt_path = /usr/local/bin/nvmf_tgt\\nrpc_socket_dir = /var/tmp/\\nrpc_socket_name = spdk.sock\\ntimeout = 60.0\\ncluster_connections = 32\\nprotocol_log_level = WARNING\\nconn_retries = 10\\ntransports = tcp\\ntransport_tcp_options = {\\"in_capsule_data_size\\": 8192, \\"max_io_qpairs_per_ctrlr\\": 7}\\nenable_dsa_acceleration = False\\nrbd_with_crc32c = True\\ntgt_cmd_extra_args = --cpumask=0xFF --msg-mempool-size=524288\\nqos_timeslice_in_usecs = 0\\nnotifications_interval = 60\\n\\n[monitor]\\ntimeout = 1.0\\n"}}}', error_ok=True, use_current_daemon_image=False)
E           Actual: _run_cephadm('test', 'nvmeof.testpool.mygroup.test.qwert', ['_orch', 'deploy'], [], stdin='{"fsid": "fsid", "name": "nvmeof.testpool.mygroup.test.qwert", "image": "", "deploy_arguments": [], "params": {"tcp_ports": [5500, 4420, 8009, 10008]}, "meta": {"service_name": "nvmeof.testpool.mygroup", "ports": [5500, 4420, 8009, 10008], "ip": null, "deployed_by": [], "rank": null, "rank_generation": null, "extra_container_args": null, "extra_entrypoint_args": null}, "config_blobs": {"config": "", "keyring": "[client.nvmeof.testpool.mygroup.test.qwert]\\nkey = None\\n", "files": {"ceph-nvmeof.conf": "# This file is generated by cephadm.\\n[gateway]\\nname = client.nvmeof.testpool.mygroup.test.qwert\\ngroup = mygroup\\naddr = 192.168.100.100\\nport = 5500\\nenable_auth = False\\nstate_update_notify = True\\nstate_update_interval_sec = 5\\nbreak_update_interval_sec = 25\\nenable_spdk_discovery_controller = False\\nencryption_key = /encryption.key\\nrebalance_period_sec = 7\\nmax_gws_in_grp = 16\\nmax_ns_to_change_lb_grp = 8\\nenable_prometheus_exporter = True\\nprometheus_exporter_ssl = False\\nprometheus_port = 10008\\nprometheus_stats_interval = 10\\nprometheus_startup_delay = 240\\nprometheus_connection_list_cache_expiration = 60\\nverify_nqns = True\\nverify_keys = True\\nverify_listener_ip = True\\n# This is a development flag, do not change it\\nabort_on_errors = True\\n# This is a development flag, do not change it\\nabort_on_update_error = True\\n# This is a development flag, do not change it\\nomap_file_ignore_unlock_errors = False\\n# This is a development flag, do not change it\\nomap_file_lock_on_read = True\\nomap_file_lock_duration = 20\\nomap_file_lock_retries = 30\\nomap_file_lock_retry_sleep_interval = 1.0\\nomap_file_update_reloads = 10\\nallowed_consecutive_spdk_ping_failures = 1\\nspdk_ping_interval_in_seconds = 2.0\\nping_spdk_under_lock = False\\nenable_monitor_client = True\\nmax_hosts_per_namespace = 16\\nmax_namespaces_with_netmask = 1000\\nmax_subsystems = 128\\nmax_hosts = 2048\\nmax_namespaces = 4096\\nmax_namespaces_per_subsystem = 512\\nmax_hosts_per_subsystem = 128\\nsubsystem_cache_expiration = 5\\nforce_tls = False\\n# This is a development flag, do not change it\\nmax_message_length_in_mb = 4\\ndegrade_namespace_on_kmip_error = True\\n\\n[gateway-logs]\\nlog_level = INFO\\nlog_files_enabled = True\\nlog_files_rotation_enabled = True\\nverbose_log_messages = True\\nmax_log_file_size_in_mb = 10\\nmax_log_files_count = 20\\nmax_log_directory_backups = 10\\nlog_directory = /var/log/ceph/\\n\\n[discovery]\\naddr = 192.168.100.100\\nport = 8009\\n# This is a development flag, do not change it\\nabort_on_errors = True\\nbind_retries_limit = 10\\nbind_sleep_interval = 0.5\\n\\n[ceph]\\npool = testpool\\nconfig_file = /etc/ceph/ceph.conf\\nid = nvmeof.testpool.mygroup.test.qwert\\n\\n[mtls]\\nserver_key = /server.key\\nclient_key = /client.key\\nserver_cert = /server.cert\\nclient_cert = /client.cert\\nroot_ca_cert = /root.ca.cert\\n\\n[kmip]\\ncert_dir = ./certs/kmip/{server_name}\\n\\n[spdk]\\ntgt_path = /usr/local/bin/nvmf_tgt\\nrpc_socket_dir = /var/tmp/\\nrpc_socket_name = spdk.sock\\ntimeout = 60.0\\ncluster_connections = 32\\nprotocol_log_level = WARNING\\nconn_retries = 10\\ntransports = tcp\\ntransport_tcp_options = {\\"in_capsule_data_size\\": 8192, \\"max_io_qpairs_per_ctrlr\\": 7}\\nenable_dsa_acceleration = False\\nrbd_with_crc32c = True\\ntgt_cmd_extra_args = --cpumask=0xFF --msg-mempool-size=524288\\nqos_timeslice_in_usecs = 0\\nnotifications_interval = 60\\n\\n[monitor]\\ntimeout = 1.0\\n"}}}', use_current_daemon_image=False, error_ok=True)
```

looks relevant

--- a/src/pybind/mgr/cephadm/templates/services/nvmeof/ceph-nvmeof.conf.j2
+++ b/src/pybind/mgr/cephadm/templates/services/nvmeof/ceph-nvmeof.conf.j2
@@ -49,6 +49,7 @@ subsystem_cache_expiration = {{ spec.subsystem_cache_expiration }}
 force_tls = {{ spec.force_tls }}
 # This is a development flag, do not change it
 max_message_length_in_mb = {{ spec.max_message_length_in_mb }}
+degrade_namespace_on_kmip_error = {{ spec.degrade_namespace_on_kmip_error }}
 
 [gateway-logs]
 log_level = {{ spec.log_level }}

--- a/src/pybind/mgr/cephadm/tests/test_services.py
+++ b/src/pybind/mgr/cephadm/tests/test_services.py
@@ -398,6 +398,7 @@ subsystem_cache_expiration = 5
 force_tls = False
 # This is a development flag, do not change it
 max_message_length_in_mb = 4
+degrade_namespace_on_kmip_error = True
 
 [gateway-logs]
 log_level = INFO

--- a/src/python-common/ceph/deployment/service_spec.py
+++ b/src/python-common/ceph/deployment/service_spec.py
@@ -1431,6 +1431,7 @@ class NvmeofServiceSpec(ServiceSpec):
                  subsystem_cache_expiration: Optional[int] = 5,
                  force_tls: Optional[bool] = False,
                  max_message_length_in_mb: Optional[int] = 4,
+                 degrade_namespace_on_kmip_error: Optional[bool] = True,
                  server_key: Optional[str] = None,
                  server_cert: Optional[str] = None,
                  client_key: Optional[str] = None,
@@ -1576,6 +1577,8 @@ class NvmeofServiceSpec(ServiceSpec):
         self.force_tls = force_tls
         #: ``max_message_length_in_mb`` max protobuf message length, in mb
         self.max_message_length_in_mb = max_message_length_in_mb
+        #: ``degrade_namespace_on_kmip_error`` on a KMIP key error in update, create a degraded ns
+        self.degrade_namespace_on_kmip_error = degrade_namespace_on_kmip_error
         #: ``allowed_consecutive_spdk_ping_failures`` # of ping failures before aborting gateway
         self.allowed_consecutive_spdk_ping_failures = allowed_consecutive_spdk_ping_failures
         #: ``spdk_ping_interval_in_seconds`` sleep interval in seconds between SPDK pings
@@ -1793,6 +1796,7 @@ class NvmeofServiceSpec(ServiceSpec):
                                    "Subsystem cache expiration period")
         verify_boolean(self.force_tls, "Force TLS")
         verify_positive_int(self.max_message_length_in_mb, "Max protocol message length")
+        verify_boolean(self.degrade_namespace_on_kmip_error, "Degrade namespace on KMIP error")
         verify_non_negative_number(self.monitor_timeout, "Monitor timeout")
         verify_non_negative_int(self.port, "Port")
         verify_non_negative_int(self.discovery_port, "Discovery port")

